src/Nixfmt/Predoc.hs

```diff
@@ -0,0 +1,195 @@
+{-# LANGUAGE DeriveFoldable     #-}
+{-# LANGUAGE DeriveFunctor      #-}
+{-# LANGUAGE FlexibleInstances  #-}
+{-# LANGUAGE OverloadedStrings  #-}
+{-# LANGUAGE StandaloneDeriving #-}
+
+-- | This module implements a layer around the prettyprinter package, making it
+-- easier to use.
+module Nixfmt.Predoc
+    ( text
+    , hcat
+    , trivia
+    , group
+    , nest
+    , softbreak
+    , break
+    , softline
+    , space
+    , hardline
+    , emptyline
+    , Doc
+    , Pretty
+    , pretty
+    , putDocW
+    , fixup
+    , flatten
+    ) where
+
+import           Data.Text                      (Text, pack)
+import qualified Data.Text.Prettyprint.Doc      as PP
+import qualified Data.Text.Prettyprint.Doc.Util as PPU
+import           Prelude                        hiding (break)
+
+data Tree a = EmptyTree
+            | Leaf a
+            | Node (Tree a) (Tree a)
+            deriving (Show, Functor)
+
+-- | Sequential Lines are reduced to a single Line by taking the maximum. This
+-- means that e.g. a Space followed by an Emptyline results in just an
+-- Emptyline.
+data Line = Softbreak
+          | Break
+          | Softline
+          | Space
+          | Hardline
+          | Emptyline
+          deriving (Show, Eq, Ord)
+
+data Predoc f
+    = Trivia Text
+    | Text Text
+    | Line Line
+    -- | Group predoc indicates either all or none of the spaces in predoc
+    -- and not in a subgroup should be converted to line breaks.
+    | Group (f (Predoc f))
+    -- | Nest n predoc indicates all line start in predoc should be indented by
+    -- n more spaces than the surroundings.
+    | Nest Int (f (Predoc f))
+
+deriving instance Show (Predoc Tree)
+deriving instance Show (Predoc [])
+
+type Doc = Tree (Predoc Tree)
+type DocList = [Predoc []]
+
+class Pretty a where
+    pretty :: a -> Doc
+
+instance Pretty Text where
+    pretty = Leaf . Text
+
+instance Pretty String where
+    pretty = Leaf . Text . pack
+
+instance Semigroup (Tree a) where
+    left <> right = Node left right
+
+instance Monoid (Tree (Predoc Tree)) where
+    mempty = EmptyTree
+
+text :: Text -> Doc
+text = Leaf . Text
+
+trivia :: Text -> Doc
+trivia = Leaf . Trivia
+
+group :: Doc -> Doc
+group = Leaf . Group
+
+nest :: Int -> Doc -> Doc
+nest level = Leaf . Nest level
+
+softbreak :: Doc
+softbreak = Leaf (Line Softbreak)
+
+break :: Doc
+break = Leaf (Line Break)
+
+softline :: Doc
+softline = Leaf (Line Softline)
+
+space :: Doc
+space = Leaf (Line Space)
+
+hardline :: Doc
+hardline = Leaf (Line Hardline)
+
+emptyline :: Doc
+emptyline = Leaf (Line Emptyline)
+
+hcat :: [Doc] -> Doc
+hcat = mconcat
```

Comment:
😕

Reply:
This is just to match the interface of the `prettyprinter` library I'm wrapping.